### PR TITLE
GemGIS - Refactoring Extract XYZ, Data Clipping, Raster Interpolation

### PR DESCRIPTION
**Pull request by AlexanderJuestel**
Refactoring functions for extracting xyz coordinates

### CODE REVIEW

**Comment by AlexanderJuestel** [comment]
@endarthur I have just pushed the latest changes. Ready for your review! :)

**Review by endarthur — APPROVED** [review]
I think that's it for this review. Code is fine, and mostly I suggest some small changes in uses of typing.
In the future, it might be interesting to use more duck typing, that is, trying first and watching for exceptions instead of checking the type of everything. This might work better for some specific parts, like dealing with inputs for pyproj's CRS objects, and what can be converted or not to an ndarray.
I also think we should look for a way to abstract most of the type checking away, as there's a lot of code and definition repetition over there, but that is definitely for the future.
Anyway, great job, Alex!

File: gemgis/raster.py [hunk 1]
```diff
@@ -31,73 +31,101 @@
 import shapely
```
[the remark this answers is not in the document]
Comment (AlexanderJuestel) [reply gemgis/raster.py:31]: I will not commit that as we already import typing further up and can just add it in the line above

File: gemgis/vector.py [hunk 2]
```diff
@@ -541,213 +571,761 @@ def extract_xy(gdf: gpd.geodataframe.GeoDataFrame,
...
+        # Checking that the bbox list only has four elements
+        if len(bbox) != 4:
+            raise ValueError('Provide minx, maxx, miny and maxy values for the bbox')
+
+        # Checking that all elements of the list are of type int or float
+        if not all(isinstance(bound, (int, float)) for bound in bbox):
+            raise TypeError('Bbox values must be of type float or int')
+
+    # Checking that the target_crs is of type string
+    if not isinstance(target_crs, (str, type(None), pyproj.crs.crs.CRS)):
```
[the remark this answers is not in the document]
Comment (AlexanderJuestel) [reply gemgis/vector.py:1019]: Yeah, let us get the rough refactoring done and smaller things can be tackeled afterwards :)

File: gemgis/vector.py [hunk 3]
```diff
@@ -541,213 +571,761 @@ def extract_xy(gdf: gpd.geodataframe.GeoDataFrame,
...
+    # Checking and setting the dtypes of the GeoDataFrame
+    gdf = set_dtype(gdf)
 
     return gdf
 
 
-# Function tested
-def interpolate_raster(gdf: gpd.geodataframe.GeoDataFrame, method: str = 'nearest', **kwargs) -> np.ndarray:
+def interpolate_raster(gdf: gpd.geodataframe.GeoDataFrame,
+                       method: str = 'nearest',
```
[the remark this answers is not in the document]
Comment (AlexanderJuestel) [reply gemgis/vector.py:1318]: Never heard of Literal before :D

File: gemgis/raster.py [hunk 4]
```diff
@@ -31,73 +31,101 @@
...
+        array (np.ndarray): Array containing the raster values
+        extent (list): List containing the values for the extent of the array (minx,maxx,miny,maxy)
+        point_x (list, np.ndarray, float, int): Object containing the x coordinates of a point or points at which the
+        array value is obtained
+        point_y (list, np.ndarray, float, int): Object containing the y coordinates of a point or points at which the
+        array value is obtained
     Return:
-        sample - float value of the raster at the provided position
+        sample (float): Value/s of the raster at the provided position/s
     """
```
[the remark this answers is not in the document]
Comment (AlexanderJuestel) [reply gemgis/raster.py:50]: Yeah, totally agree :)

File: gemgis/raster.py [hunk 5]
```diff
@@ -31,73 +31,101 @@
...
+    # Checking if the point is located within the provided extent
+    if isinstance(point_x, (float, int)):
+        if point_x < extent[0] or point_x > extent[1]:
+            raise ValueError('One or multiple points are located outside of the extent')
+    if isinstance(point_y, (float, int)):
+        if point_y < extent[2] or point_y > extent[3]:
+            raise ValueError('One or multiple points are located outside of the extent')
+
+    # Converting lists of coordinates to np.ndarrays
+    if isinstance(point_x, list) and isinstance(point_y, list):
```
[the remark this answers is not in the document]
Comment (AlexanderJuestel) [reply gemgis/raster.py:107]: Could you open an issue for that? Just that we do not forget it and can integrate it in a later PR? :)

File: gemgis/raster.py [hunk 6]
```diff
@@ -31,132 +31,277 @@
...
+        sample_outside_extent (bool): Allow sampling outside the extent of the rasterio object, default is True
     Return:
-        tuple - float of sampled raster value and list containing the x- and y-coordinates of the point where the
-        sample was drawn
+        sample (list, float): Value/s of the raster at the provided position/s
     """
 
-    seed = kwargs.get('seed', None)
+    # Checking that the raster is a rasterio object
+    if not isinstance(raster, rasterio.io.DatasetReader):
```
[the remark this answers is not in the document]
Comment (AlexanderJuestel) [reply gemgis/raster.py:154]: Could be a nice addition for the future :)

File: gemgis/raster.py [hunk 7]
```diff
@@ -31,132 +31,277 @@
...
-    if isinstance(array, rasterio.io.DatasetReader):
-        array = array.read(1)
+    # Checking that n is of type int
+    if not isinstance(n, int):
+        raise TypeError('Number of samples n must be provided as int')
 
     # Checking if extent is a list
-    if not isinstance(extent, list):
+    if not isinstance(seed, (int, type(None))):
         raise TypeError('Extent must be of type list')
```
[the remark this answers is not in the document]
Comment (AlexanderJuestel) [reply gemgis/raster.py:251]: Ups :D

**Pull request merged**